firedrake/matrix_free/preconditioners.py

```diff
@@ -0,0 +1,314 @@
+
+        V = test.function_space()
+
+        u = TrialFunction(V)
+        v = TestFunction(V)
+        # Handle vector and tensor-valued spaces.
+        a = inner(u, v)*dx
+
+        A = assemble(a, form_compiler_parameters=context.fc_params)
+        A.force_evaluation()
+
+        Pmat = A.petscmat
+        Pmat.setNullSpace(P.getNullSpace())
+        Pmat.setTransposeNullSpace(P.getTransposeNullSpace())
+
+        ksp = PETSc.KSP().create()
+        ksp.setOperators(Pmat)
+        ksp.setOptionsPrefix(prefix + "Mp_")
+        ksp.setUp()
+        ksp.setFromOptions()
+        self.ksp = ksp
+
+    def update(self, pc):
+        pass
+
+    def apply(self, pc, X, Y):
+        self.ksp.solve(X, Y)
+
+    # Mass matrix is symmetric
+    applyTranspose = apply
+
+    def view(self, pc, viewer=None):
+        super(MassInvPC, self).view(pc, viewer)
+        viewer.printfASCII("KSP solver for M^-1\n")
+        viewer.pushASCIITab()
+        self.ksp.view(viewer)
+        viewer.popASCIITab()
+
+
+class PCDPC(PCBase):
+    """A Pressure-Convection-Diffusion preconditioner for Navier-Stokes.
+
+    This preconditioner approximates the inverse of the pressure schur
+    complement for the Navier-Stokes equations by.
+
+    .. math::
+
+       S^{-1} \approx K^{-1} F_p M^{-1}
+
+    Where :math:`K = \nabla^2`, :math:`M = \mathbb{I}` and
+    :math:`F_p = 1/\mathrm{Re} \nabla^2 + u\cdot\grad`.
+
+    The inverse of :math:`K` is approximated by a KSP which can be
+    controlled using the options prefix ``pcd_Kp_``.
+
+    The inverse of :math:`M` is similarly approximated by a KSP which
+    can be controlled using the options prefix ``pcd_Mp_``.
+
+    :math:`F_p` requires both the Reynolds number and the current
+    velocity.  You must provide these with options using the glbaol
+    option ``Re`` for the Reynolds number and the prefixed option
+    ``pcd_velocity_space`` which should be the index into the full
+    space that gives velocity field.
+
+    """
+    def initialize(self, pc):
+        from firedrake import TrialFunction, TestFunction, dx, \
+            assemble, inner, grad, split, Constant
+        prefix = pc.getOptionsPrefix() + "pcd_"
+
+        # we assume P has things stuffed inside of it
+        _, P = pc.getOperators()
+        context = P.getPythonContext()
+
+        test, trial = context.a.arguments()
+        if test.function_space() != trial.function_space():
+            raise ValueError("Pressure space test and trial space differ")
+
+        Q = test.function_space()
+
+        p = TrialFunction(Q)
+        q = TestFunction(Q)
+
+        mass = p*q*dx
+
+        # Regularisation to avoid having to think about nullspaces.
+        stiffness = inner(grad(p), grad(q))*dx + Constant(1e-6)*p*q*dx
+
+        # FIXME: allow these guys to be matrix-free
+        Mp = assemble(mass, form_compiler_parameters=context.fc_params)
+        Kp = assemble(stiffness, form_compiler_parameters=context.fc_params)
```

Comment:
These are scalar mass and stiffness matrices.  Nesting doesn't make sense.
But it's conceivable that one might want
matrix-free (e.g. if we're in a high-order discretization and we want to do
Schwarz or some other Python-level preconditioner.)
I'll look at this.


On Fri, Aug 12, 2016 at 11:08 AM, Lawrence Mitchell <
notifications@github.com> wrote:

> In firedrake/matrix_free/preconditioners.py
> <https://github.com/firedrakeproject/firedrake/pull/853#discussion_r74617246>
> :
>
> > +        if test.function_space() != trial.function_space():
> > +            raise ValueError("Pressure space test and trial space differ")
> > +
> > +        Q = test.function_space()
> > +
> > +        p = TrialFunction(Q)
> > +        q = TestFunction(Q)
> > +
> > +        mass = p*q*dx
> > +
> > +        # Regularisation to avoid having to think about nullspaces.
> > +        stiffness = inner(grad(p), grad(q))*dx + Constant(1e-6)*p*q*dx
> > +
> > +        # FIXME: allow these guys to be matrix-free
> > +        Mp = assemble(mass, form_compiler_parameters=context.fc_params)
> > +        Kp = assemble(stiffness, form_compiler_parameters=context.fc_params)
>
> Likewise options nest=True/False/matfree.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/firedrakeproject/firedrake/pull/853/files/e2b7431b503c5fb92a59a90c62cbbb8feb8ef0ba#r74617246>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADl6L2eShPLC1Ac_w-_RydIfrN9xnSGeks5qfJp8gaJpZM4JiV-g>
> .
>